### PR TITLE
F/cluster function split

### PR DESCRIPTION
**Pull request by tazitoo**
Updated:
1) clustering API - allows users to use external routine
2) internal clustering routines: build & fastPAM for exhaustive search on smaller datasets, banditPAM for approximate search on large datasets, scales like O(nlogn)
3) some clustering example scripts - useful test cases, might be useful for longer running tests

### CODE REVIEW

**Review by rreilly10 — CHANGES_REQUESTED** [review]
This looks really great - straight forward and easy to understand. I do have a few requests and thoughts that I added in the comments below.

I had to install `numba` - we should put that in the `requirements.txt`
Looks like`numba==0.51.2` will work

Also, if you could install flake8 and run the `lint_local` command that would be great.
Looks like quite a few new linting issues were introduced that are breaking the pipeline.

```
pip install flake8
make lint_local
```

If there are any linting issues you see that you want to change we can also add those to the `.flake8` file

I'll submit a PR soon that will include all of the testing/linting dependencies, and a better contribution guide for what to run locally vs. what is run when a PR is opened

File: tests/test_clustering.py [hunk 1]
```diff
@@ -5,8 +5,8 @@
```
[the remark this answers is not in the document]
Comment (tazitoo) [reply tests/test_clustering.py:5]: created a test from one of the clustering examples

File: examples/clustering/cluster_700_9.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+kmed2.fit(attributions, verbose=True)
+end_time = time.time()
+elapsed_time = end_time - start_time
+print(f"Finished test in {elapsed_time:.2f}")
+print(kmed2.centers)
+# test that 2 attributions are in each cluster
+# assert(sum(kmedoids_2.members) == 2)
+
+
+#pca = PCA(n_components=2)
```
[the remark this answers is not in the document]
Comment (tazitoo) [reply examples/clustering/cluster_700_9.py]: I think I cleaned up most of the commented code...

**Review by tazitoo — COMMENTED** [review]
I think all outstanding issues were cleaned up...

File: tests/test_banditPAM.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    )
+    start_time = time.time()
+    kmed2.fit(attributions, verbose=False)
+    end_time = time.time()
+    elapsed_time = end_time - start_time
+    print(f"Finished test in {elapsed_time:.2f}")
+    print(kmed2.centers)
+
+
+    # if testing with 'euclidean' distance
```
Comment (rreilly10) [inline tests/test_banditPAM.py:36]: We could just create two tests to handle both of these. One for euclidian and one for spearman

**Review by rreilly10 — CHANGES_REQUESTED** [review]
Looks great - just one more request for separate euclidian and spearman tests

**Review by rreilly10 — APPROVED** [review]

**Pull request merged**